### PR TITLE
Backports v0.18 PR:14310, 14305 and 14307 with dash specific changes

### PR DESCRIPTION
**Pull request by pravblockc**
Backports v0.18 PR:14310, 14305 and 14307 with dash specific changes

### CODE REVIEW

File: test/functional/test_framework/messages.py [hunk 1]
```diff
@@ -428,7 +443,10 @@ def __repr__(self):
                   bytes_to_hex_str(self.scriptPubKey))
 
 
-class CTransaction():
+class CTransaction:
+    __slots__ = ("hash", "nLockTime", "nVersion", "sha256", "vin", "vout",
+                 "wit", "nType", "vExtraPayload")
```
Comment (UdjinM6) [inline test/functional/test_framework/messages.py:448]: 14305:
```suggestion
                 "nType", "vExtraPayload")
```

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
14305: should add `__slots__` to Dash-specific classes/messages too

File: test/functional/test_framework/messages.py (hunk 1 above)
Comment (pravblockc) [reply test/functional/test_framework/messages.py]: The above mentioned changes applied, Thank you

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
Pls see https://github.com/UdjinM6/dash/commit/7250a59bc39742a86e7f6b31d3b63d24a7b12083

**Comment by pravblockc** [comment]
> Pls see [UdjinM6@7250a59](https://github.com/UdjinM6/dash/commit/7250a59bc39742a86e7f6b31d3b63d24a7b12083)

Thank you

**Comment by pravblockc** [comment]
Changes has been squashed

**Review by UdjinM6 — APPROVED** [review]
Looks good, utACK

File: test/functional/test_framework/messages.py [hunk 2]
```diff
@@ -10,7 +10,12 @@
     bitcoin/primitives
 msg_block, msg_tx, msg_headers, etc.:
     data structures that represent network messages
-ser_*, deser_*: functions that handle serialization/deserialization."""
```
Comment (PastaPastaPasta) [inline test/functional/test_framework/messages.py:13]: 14305 added p2p_segwit.py

File: test/functional/test_framework/messages.py [hunk 3]
```diff
@@ -428,7 +444,10 @@ def __repr__(self):
                   bytes_to_hex_str(self.scriptPubKey))
 
 
-class CTransaction():
+class CTransaction:
+    __slots__ = ("hash", "nLockTime", "nVersion", "sha256", "vin", "vout",
+                 "wit")
```
Comment (PastaPastaPasta) [inline test/functional/test_framework/messages.py]: Do we need to / should we remove 'wit'?

File: test/functional/test_framework/messages.py [hunk 4]
```diff
@@ -680,9 +711,12 @@ def calculate_shortid(k0, k1, tx_hash):
     expected_shortid &= 0x0000ffffffffffff
     return expected_shortid
 
+
 # This version gets rid of the array lengths, and reinterprets the differential
 # encoding into indices that can be used for lookup.
-class HeaderAndShortIDs():
+class HeaderAndShortIDs:
+    __slots__ = ("header", "nonce", "prefilled_txn", "shortids", "use_witness")
```
Comment (PastaPastaPasta) [inline test/functional/test_framework/messages.py:718]: wit

File: test/functional/test_framework/messages.py [hunk 5]
```diff
@@ -1244,7 +1288,8 @@ def __repr__(self):
         return "msg_inv(inv=%s)" % (repr(self.inv))
 
 
-class msg_getdata():
+class msg_getdata:
+    __slots__ = ("inv",)
```
Comment (PastaPastaPasta) [inline test/functional/test_framework/messages.py:1313]: need to modify https://github.com/dashpay/dash/blob/6e0fdc7321e399ec2fdc2874bb930b4cc956348d/test/functional/test_framework/messages.py#L1228-L1229

File: test/functional/test_framework/messages.py [hunk 6]
```diff
@@ -1330,7 +1381,8 @@ def serialize(self):
     def __repr__(self):
         return "msg_generic()"
 
-class msg_getaddr():
+class msg_getaddr:
```
Comment (PastaPastaPasta) [inline test/functional/test_framework/messages.py:1404]: add newline above this

**Review by PastaPastaPasta — CHANGES_REQUESTED** [review]
All the fixes in 14307 shouldn't be :)

File: test/functional/test_framework/messages.py (hunk 2 above)
Comment (pravblockc) [reply test/functional/test_framework/messages.py:13]: resolved

File: test/functional/test_framework/messages.py (hunk 5 above)
Comment (pravblockc) [reply test/functional/test_framework/messages.py:1313]: fixed in different commit and brought in with edit

File: test/functional/test_framework/messages.py (hunk 6 above)
Comment (pravblockc) [reply test/functional/test_framework/messages.py:1405]: done

File: test/functional/test_framework/messages.py (hunk 3 above)
Comment (pravblockc) [reply test/functional/test_framework/messages.py]: resolved in different commit and brought in to this commit, fixed

**Review by PastaPastaPasta — APPROVED** [review]
LGTM, utACK for merge via merge commit

**Review by UdjinM6 — APPROVED** [review]
utACK

**Pull request merged**